<?xml version="1.0"?>
<robot
    xmlns:xacro="http://www.ros.org/wiki/xacro" name="ninja_link" >

  <xacro:macro name="ninja_center_link" params="child1 child2 child3">
    ## dummy
    <link name="root">
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.00001"/>
        <inertia
            ixx="0.000001" ixy="0.0" ixz="0.0"
            iyy="0.000001" iyz="0.0"
            izz="0.000001"/>
      </inertial>
    </link>
    <joint name="root_joint" type="fixed">
      <parent link="root"/>
      <child link="main_body"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <axis xyz="0 0 1"/>
    </joint>

    ## link
    <link name="main_body">
      <inertial>
        <origin xyz="-0.08989427 -0.00260694 -0.01604721" rpy="0 0 0"/>
        <mass value="1.32590383"/>
        <inertia
            ixx = "0.00565938" ixy = "0.00127382" ixz = "0.00065740"
            iyx = "0.00127382" iyy = "0.04578657" iyz = "-0.00004725"
            izx = "0.00065q740" izy = "-0.00004725" izz = "0.04949468"/>
      </inertial>
        <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <mesh filename="package://ninja/urdf/mesh/mainbody.dae" scale="1 1 1"/>
        </geometry>
      </visual>
      <collision>
        <origin xyz="-0.12 0 -0.052" rpy="0 0 0"/>
        <geometry>
          <box size="0.30 0.3065 0.104"/>
        </geometry>
      </collision>
    </link>

    <xacro:damping_factor link="main_body"/>

    <link name="fc">
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.00001"/>
        <inertia
            ixx="0.000001" ixy="0.0" ixz="0.0"
            iyy="0.000001" iyz="0.0"
            izz="0.000001"/>
      </inertial>
    </link>
    <joint name="fc_joint" type="fixed">
      <child link="fc"/>
      <parent link="main_body"/>
      <origin rpy="0 0 0" xyz="-0.04695 0 0.0369"/>
      <axis xyz="0 0 1"/>
    </joint>

    ## dummy rotor arms
    <link name="rotor_arm${child1}">
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.001"/>
        <inertia
            ixx="0.000001" ixy="0.0" ixz="0.0"
            iyy="0.000001" iyz="0.0"
            izz="0.000001"/>
      </inertial>
    </link>
    <joint name="base_arm_joint${child1}" type="fixed">
      <parent link="main_body"/>
      <child link="rotor_arm${child1}"/>
      <origin rpy="0 0 ${pi/2}" xyz="0.008 0.02725 -0.0165"/>
      <axis xyz="0 0 1"/>
    </joint>

    <link name="rotor_arm${child2}">
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.0001"/>
        <inertia
            ixx="0.000001" ixy="0.0" ixz="0.0"
            iyy="0.000001" iyz="0.0"
            izz="0.000001"/>
      </inertial>
    </link>
    <joint name="base_arm_joint${child2}" type="fixed">
      <parent link="main_body"/>
      <child link="rotor_arm${child2}"/>
      <origin rpy="0 0 ${-pi/2}" xyz="0.008 -0.02725 -0.0165"/>
      <axis xyz="0 0 1"/>
    </joint>

    <link name="rotor_arm${child3}">
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.0001"/>
        <inertia
            ixx="0.000001" ixy="0.0" ixz="0.0"
            iyy="0.000001" iyz="0.0"
            izz="0.000001"/>
      </inertial>
    </link>
    <joint name="base_arm_joint${child3}" type="fixed">
      <parent link="main_body"/>
      <child link="rotor_arm${child3}"/>
      <origin rpy="0 0 ${pi}" xyz="-0.0625 0 -0.0165"/>
      <axis xyz="0 0 1"/>
    </joint>

   <xacro:friction self="main_body"/>
  </xacro:macro>

  <xacro:macro name="ninja_docking_mech_yaw">
    <!-- yaw -->
    <link name="yaw_dock_link">
      <inertial>
        <origin xyz="0.06781112 -0.00002493 -0.02345044" rpy="0 0 0"/>
        <mass value="0.21443253"/>
        <inertia
            ixx = "0.00093162" ixy = "0.00000022" ixz = "-0.00002253"
            iyx = "0.00000022" iyy = "0.00071768" iyz = "-0.00010593"
            izx = "-0.00002253" izy = "-0.00010593" izz = "0.00091594"/>
      </inertial>
      <visual>
        <origin xyz="0 0 0" rpy="${pi/2} 0 0"/>
        <geometry>
          <mesh filename="package://ninja/urdf/mesh/female_dock_mech.dae" scale="1 1 1"/>
        </geometry>
      </visual>
      <!-- <collision> -->
      <!--   <origin xyz="0 0 0" rpy="0 0 0"/> -->
      <!--   <geometry> -->
      <!--     <box size="0.05 0.05 0.05"/> -->
      <!--   </geometry> -->
      <!-- </collision> -->
    </link>
    <xacro:damping_factor link="yaw_dock_link"/>

    <joint name="yaw_dock_joint" type="revolute">
      <limit effort="6.6" lower="-3.14" upper="3.14" velocity="0.2"/>
      <parent link="main_body"/>
      <child link="yaw_dock_link"/>
      <origin rpy="0 0 0" xyz="0.1935 0 0.0085"/>
      <axis xyz="0 0 1"/>
      <dynamics damping="0.8" friction="0.0"/>
    </joint>

    <transmission name="yaw_dockjoint_tran">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="yaw_dock_joint">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
      <actuator name="yaw_dock_servo">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>

    <link name="yaw_connect_point">
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.0001"/>
        <inertia
            ixx="0.000001" ixy="0.0" ixz="0.0"
            iyy="0.000001" iyz="0.0"
            izz="0.000001"/>
      </inertial>
    </link>
    <joint name="yaw_connect_point_joint" type="fixed">
      <parent link="yaw_dock_link"/>
      <child link="yaw_connect_point"/>
      <origin rpy="0 0 0" xyz="0.037 0 -0.025"/>
      <axis xyz="0 0 1"/>
    </joint>
    
  </xacro:macro>

  <xacro:macro name="ninja_gimbal_link" params="self rotor_direction">
    #### gimbal joints ####
    <joint name="gimbal${self}_roll" type="revolute">
      <limit effort="6.6" lower="-2.0" upper="2.0" velocity="3.0"/>
      <parent link="rotor_arm${self}"/>
      <child link="gimbal_link${self}"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <axis xyz="1 0 0"/>
      <dynamics damping="0.8" friction="0.0"/>
    </joint>

    <transmission name="joint_tran${2*self-1}">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="gimbal${self}_roll">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
      <actuator name="servo${2*self-1}">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>

    #### gimbal link  ####
    <link name="gimbal_link${self}">
      <inertial>
       <origin xyz="0.12245044 0.03129270 0.00001719" rpy="0 0 0"/>
        <mass value="0.19445396"/>
        <inertia
            ixx = "0.00119253" ixy = "0.00015756" ixz = "-0.00000029"
            iyx = "0.00015756" iyy = "0.00043864" iyz = "-0.00000018"
            izx = "-0.00000029" izy = "-0.00000018" izz = "0.00156562"/>
      </inertial>
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <mesh filename="package://ninja/urdf/mesh/gimbal_link.dae" scale="1 1 1"/>
        </geometry>
      </visual>
    </link>
    <xacro:damping_factor link="gimbal_link${self}"/>


    #### rotor joint  ####
    <joint name="gimbal${self}_pitch" type="revolute">
      <limit effort="6.6" lower="-2.0" upper="2.0" velocity="3.0"/>
      <parent link="gimbal_link${self}"/>
      <child link="rotor_parent${self}"/>
      <origin rpy="0 0 0" xyz="0.162845 0.09285 0"/>
      <axis xyz="0 1 0"/>
    </joint>
    <transmission name="joint_tran${2*self}">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="gimbal${self}_pitch">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
      <actuator name="servo${2*self}">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>

    ## rotor parent link
    <link name="rotor_parent${self}">
      <inertial>
        <origin xyz="0.09 0 0" rpy="0 0 ${-pi/2}"/>
        <mass value="0.220"/>
        <inertia
            ixx="0.0002011" ixy="0.0000001" ixz="0.0000003"
            iyy="0.0004337" iyz="0.0"
            izz="0.0005261"/>
      </inertial>
      <visual>
        <origin xyz="0 -0.09285 0" rpy="0 0 ${pi}"/>
        <geometry>
          <mesh filename="package://ninja/urdf/mesh/5inch_edf.dae" scale="1 1 1"/>
        </geometry>
      </visual>
    </link>

    ## thrust link
    <link name="thrust${self}">
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.00001"/>
        <inertia
            ixx="0.001" ixy="0.0" ixz="0.0"
            iyy="0.001" iyz="0.0"
            izz="0.001"/>
      </inertial>
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <cylinder radius="0.1" length="0.1"/>
        </geometry>
      </collision>
    </link>
    <xacro:damping_factor link="thrust${self}"/>

    #### rotor joint (virtual) ####
    <joint name="rotor${self}" type="continuous">
      <limit effort="100.0" lower="${min_force}" upper="${max_force}" velocity="0.5"/>
      <parent link="rotor_parent${self}"/>
      <child link="thrust${self}"/>
      <origin rpy="0 0 0" xyz="0 -0.09285 0.025"/>
      <axis xyz="0 0 ${rotor_direction}"/>
    </joint>

    <transmission name="rotor_tran${self}">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="rotor${self}">
        <hardwareInterface>RotorInterface</hardwareInterface>
      </joint>
      <actuator name="rotor_actuator${self}">
        <hardwareInterface>RotorInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>

  </xacro:macro>

  <xacro:macro name="ninja_docking_mech_pitch">
    <!-- pitch -->
    <link name="pitch_dock_link">
      <inertial>
        <origin xyz="-0.06961517 0.02389743 -0.00688546" rpy="0 0 0"/>
        <mass value="0.30174800"/>
        <inertia
            ixx = "0.00119253" ixy = "0.00015756" ixz = "-0.00000029"
            iyx = "0.00015756" iyy = "0.00043864" iyz = "-0.00000018"
            izx = "-0.00000029" izy = "-0.00000018" izz = "0.00156562"/>
      </inertial>
      <visual>
        <origin xyz="0 0 0" rpy="${pi} ${pi} 0"/>
        <geometry>
          <mesh filename="package://ninja/urdf/mesh/male_dock_mech.dae" scale="1 1 1"/>
        </geometry>
      </visual>
      <collision>
        <origin xyz="-${0.1531/2} 0.025 0" rpy="0 0 0"/>
        <geometry>
          <box size="0.1531 0.064 0.064"/>
        </geometry>
      </collision>
    </link>
    <xacro:damping_factor link="pitch_dock_link"/>

    <joint name="pitch_dock_joint" type="revolute">
      <limit effort="6.6" lower="-3.14" upper="3.14" velocity="0.2"/>
      <parent link="main_body"/>
      <child link="pitch_dock_link"/>
      <origin rpy="0 0 0" xyz="-0.434658 -0.025 -0.0165"/>
      <axis xyz="0 1 0"/>
      <dynamics damping="0.8" friction="0.0"/>
    </joint>

    <transmission name="pitch_dockjoint_tran">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="pitch_dock_joint">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
      <actuator name="pitch_dock_servo">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>

    <link name="pitch_connect_point">
      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.0001"/>
        <inertia
            ixx="0.000001" ixy="0.0" ixz="0.0"
            iyy="0.000001" iyz="0.0"
            izz="0.000001"/>
      </inertial>
    </link>
    <joint name="pitch_connect_point_joint" type="fixed">
      <parent link="pitch_dock_link"/>
      <child link="pitch_connect_point"/>
      <origin rpy="0 0 0" xyz="-0.1531 0.025 0"/>
      <axis xyz="0 0 1"/>
    </joint>

  </xacro:macro>
  
  <xacro:macro name="battery">
    <joint name="battry_1_joint" type="fixed">
      <parent link="main_body"/>
      <child link="battery_1"/>
      <origin rpy="-${pi/2} 0 ${pi}" xyz="-0.00891 0.0399 -0.069"/>
      <axis xyz="0 0 0"/>
    </joint>

    <link name="battery_1">
      <inertial>
       <origin xyz="0.12 0.029 0.0000178" rpy="0 0 0"/>
        <mass value="0.18778"/>
        <inertia
            ixx="0.00119072" iyy="0.00041760" izz="0.00154581"
            ixy="0.00012643" ixz=" -0.00000030" iyz="-0.00000017"/>
      </inertial>
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <mesh filename="package://ninja/urdf/mesh/battery.dae" scale="1 1 1"/>
        </geometry>
      </visual>
    </link>
    <xacro:damping_factor link="battery_1"/>

    <joint name="battry_2_joint" type="fixed">
      <parent link="main_body"/>
      <child link="battery_2"/>
      <origin rpy="${-pi/2} 0 0" xyz="-0.00891 -0.0399 -0.069"/>
      <axis xyz="0 0 0"/>
    </joint>

    <link name="battery_2">
      <inertial>
       <origin xyz="0 0.0 00.01700000" rpy="0 0 0"/>
        <mass value="0.40700000"/>
        <inertia
            ixx="0.00010170" iyy="0.00060259" izz="0.00062827"
            ixy="0.000" ixz=" 0.000000" iyz="0.0000"/>
      </inertial>
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <mesh filename="package://ninja/urdf/mesh/battery.dae" scale="1 1 1"/>
        </geometry>
      </visual>
    </link>
    <xacro:damping_factor link="battery_2"/>

  </xacro:macro>
  
</robot>
